src/clib/pio_msg.c

```diff
@@ -0,0 +1,2179 @@
+    int varid;
+    int namelen, newnamelen;
+    char *name, *newname;
+    int mpierr;
+    int ret;
+
+    LOG((1, "delete_att_handler"));
+
+    /* Get the parameters for this function that the he comp master
+     * task is broadcasting. */
+    if ((mpierr = MPI_Bcast(&ncid, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&varid, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&namelen, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if (!(name = malloc((namelen + 1) * sizeof(char))))
+	return PIO_ENOMEM;
+    if ((mpierr = MPI_Bcast(name, namelen + 1, MPI_CHAR, 0, ios->intercomm)))
+	return PIO_EIO;
+    LOG((2, "delete_att_handler namelen = %d name = %s ncid = %d varid = %d ",
+	 namelen, name, ncid, varid));
+
+    /* Call the create file function. */
+    if ((ret = PIOc_del_att(ncid, varid, name)))
+	return ret;
+
+    /* Free resources. */
+    free(name);
+
+    LOG((1, "delete_att_handler succeeded!"));
+    return PIO_NOERR;
+}
+
+/** This function is run on the IO tasks. It reads or writes an array
+ *  of data to a netCDF variable.
+ *
+ * @param ios pointer to the iosystem_desc_t data.
+ * @param msg the message sent my the comp root task.
+ *
+ * @return PIO_NOERR for success, error code otherwise. */
+int vara_handler(iosystem_desc_t *ios, int msg)
+{
+    int ncid;
+    int len, namelen;
+    int iotype;
+    char *name;
+    int mode;
+    int mpierr;
+    int ret;
+    int varid;
+    nc_type xtype;
+    int ndims;
+    int *dimids;
+    void *data;
+    int size_in_bytes;
+    PIO_Offset *count, *start;
+
+    int my_rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
+    LOG((1, "%d def_var_handler comproot = %d\n", my_rank, ios->comproot));
+
+    if (msg == PIO_MSG_PUT_VARA)
+    {
+	/* Get the parameters for this function that the he comp master
+	 * task is broadcasting. */
+	if ((mpierr = MPI_Bcast(&ncid, 1, MPI_INT, 0, ios->intercomm)))
+	    return PIO_EIO;
+	if ((mpierr = MPI_Bcast(&varid, 1, MPI_INT, 0, ios->intercomm)))
+	    return PIO_EIO;
+	if ((ret = PIOc_inq_varndims(ncid, varid, &ndims)))
+	    return ret;
+	if (!(start = malloc(ndims * sizeof(int))))
+	    return PIO_ENOMEM;
+	if (!(count = malloc(ndims * sizeof(int))))
+	    return PIO_ENOMEM;
+	if ((mpierr = MPI_Bcast(start, ndims, MPI_INT, 0, ios->intercomm)))
+	    return PIO_EIO;
+	if ((mpierr = MPI_Bcast(count, ndims, MPI_INT, 0, ios->intercomm)))
+	    return PIO_EIO;
+	int size = 1;
+	for (int d = 0; d < ndims; d++)
+	    size *= count[d];
+	size_in_bytes = size * sizeof(int);
+	if (!(data = malloc(size_in_bytes)))
+	    return PIO_ENOMEM;
+	if ((mpierr = MPI_Bcast(data, size_in_bytes, MPI_INT, 0,
+				ios->intercomm)))
+	    return PIO_EIO;
+	LOG((2, " def_var_handler got parameters namelen = %d "
+	     "name = %s len = %d ncid = %d", namelen, name, len, ncid));
+
+	/* Call the create file function. */
+	if ((ret = PIOc_put_vara_int(ncid, varid, start, count, data)))
```

Comment:
This function should have been (and now is) removed. It was an artifact of an earlier attempt in which I intended to use one handler for puts and gets. But the code between the two was mostly different so I use two separate functions now.